### PR TITLE
Strength reduction should not remove some casts

### PR DESCRIPTION
**Pull request by mbudiu-vmw**
Signed-off-by: Mihai Budiu <mbudiu@vmware.com>

### CODE REVIEW

**Comment by mbudiu-vmw** [comment]
Fixed #3225

File: frontends/p4/strengthReduction.cpp [hunk 1]
```diff
@@ -444,10 +444,22 @@ const IR::Node* DoStrengthReduction::postorder(IR::Slice* expr) {
     }
 
     while (auto cast = expr->e0->to<IR::Cast>()) {
-        if (expr->getH() < size_t(cast->expr->type->width_bits())) {
-            expr->e0 = cast->expr;
+        if (auto tb = cast->expr->type->to<IR::Type_Bits>()) {
+            auto type = expr->type->to<IR::Type_Bits>();
```
Comment (fruffy) [inline frontends/p4/strengthReduction.cpp:448]: Why not use checkedTo here?

File: frontends/p4/strengthReduction.cpp [hunk 2]
```diff
@@ -444,10 +444,22 @@ const IR::Node* DoStrengthReduction::postorder(IR::Slice* expr) {
...
+            CHECK_NULL(type);
+            if (tb->isSigned != type->isSigned)
```
Comment (fruffy) [inline frontends/p4/strengthReduction.cpp:450]: I always assumed that slices of signed bit strings will produce an unsigned bit string. This should probably ensure that a cast to `int<W>` is converted into a `bit<W>`. A cast from `int<W>` to bit<W> should be acceptable though? 

This is what the spec says: 
>The result is an unsigned bit-string of width m - l + 1, including the bits numbered from l (which becomes the least significant bit of the result) to m (the most significant bit of the result) from the source operand.

Comment (mbudiu-vmw) [reply frontends/p4/strengthReduction.cpp:450]: I added another test case which fails without this change; it's in the test file:
`h.h.b = ((bit<1>)h.h.c)[0:0];`

Comment (fruffy) [reply frontends/p4/strengthReduction.cpp:449]: But
`((bit<1>)h.h.c)[0:0];`
to
`((bit<1>)h.h.c);` 
should be acceptable, right?

**Review by fruffy — APPROVED** [review]

File: frontends/p4/strengthReduction.cpp (hunk 2 above)
Comment (mbudiu-vmw) [reply frontends/p4/strengthReduction.cpp:449]: The previous result was just h.h.c, which failed to typecheck.

**Comment by fruffy** [comment]
The only thing is to to add a new member d to the header and assign c to that one. Otherwise, the assignment is optimized away.

**Comment by fruffy** [comment]
Is this ready to be merged?

**Pull request merged**